mono/metadata/icall.c

```diff
@@ -3548,18 +3548,20 @@ ves_icall_InternalInvoke (MonoReflectionMethod *method, MonoObject *this_arg, Mo
 		}
 	}
 
-	if (sig->ret->byref) {
 #if ENABLE_NETCORE
+	if (sig->ret->byref || m_class_is_byreflike (mono_class_from_mono_type_internal (sig->ret))) {
 		MonoType* ret_byval = m_class_get_byval_arg (mono_class_from_mono_type_internal (sig->ret));
-		if (ret_byval->byref) {
+		if (ret_byval->byref || m_class_is_byreflike (mono_class_from_mono_type_internal (ret_byval))) {
 			mono_gc_wbarrier_generic_store_internal (exc, (MonoObject*) mono_exception_from_name_msg (mono_defaults.corlib, "System", "NotSupportedException", "Cannot invoke method returning ByRef to ByRefLike type via reflection"));
 			return NULL;
 		}
+	}
 #else
+	if (sig->ret->byref) {
```

Comment:
We check for ByRefLike for classic Mono, too.

```csharp
using System;

public ref struct Q {
	public int i;
}

public class Foo {
	static int v;

	public static void Main () {
		var mi = typeof(Foo).GetMethod ("Test");

		var o = mi.Invoke (null, null);
	}

	public static Q Test ()
	{
		Q q;
		q.i = 42;
		Console.WriteLine ("a");
		return q;
	}
}
```
(throws `BadImageFormatException`)

Reply:
The behaviour was undefined under NetFX so I thought it is fine to leave it as is. I can change it to return the same exception if desired.

Reply:
To expand on my previous answer... NetFX didn't support `ref struct`s in the first place. Imposing the NetFX check on Mono could break existing code that was using `Span`s with reflection which happened to be working. We can either leave the code as-is and depend on the checks lower down the stack or we can enforce the full CoreFX behaviour even on regular Mono (which is more lenient than NetFX).